### PR TITLE
Add base & implementation solidity contracts to demo secure randomness

### PR DESCRIPTION
**Pull request by sisyphusSmiling**
Closes: #17 

### Description

- Adds a Solidity implementation of CoinToss, extending a proposed secure random consumer abstract contract
- Adds Cadence transactions to allow a COA to interact with the Solidity CoinToss implementation
- Updates Cadence dependencies to use Mainnet deployments
- Updates the README to include information about EVM implementations

### CODE REVIEW

File: solidity/src/CadenceRandomConsumer.sol [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        Request memory request = _requests[requestIndex];
+
+        // Ensure that the request is fulfilled at a Flow block height greater than the one at which the request was made
+        uint64 flowHeight = _flowBlockHeight();
+        require(request.flowHeight < flowHeight, "Cannot fulfill request until subsequent Flow network block height");
+
+        // Get the random source for the Flow block at which the request was made
+        uint64 randomResult = _getRandomSource(request.flowHeight); // returns bytes32
+        // Pack the randomResult into a uint64, hashing with the requestId to vary results across shared block heights
+        randomResult = uint64(uint256(keccak256(abi.encodePacked(randomResult, requestId))));
```
Comment (tarakby) [inline solidity/src/CadenceRandomConsumer.sol:85]: That's where the PRG implementation detail can be added. We should add a comment to mention that the random source is only 32 bytes. Here only 8 bytes were required, but if the contract requires more random bytes, the source should be expanded into any arbitrary number of bytes using a PRG (which I believe is that we did on the Cadence example).

**Review by tarakby — APPROVED** [review]
Great docs and nice example 👏🏼 

I've added minor comments about the mentions of PRG. I noticed from other discussions that PRGs and their purpose have been causing confusion so I suggest to only mention them when needed (their purpose being to expand a small source into many bytes)

**Review by ramtinms — APPROVED** [review]
nice 👏

File: README.md [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
-  `revertibleRandom` is only suitable for smart contract functions that exclusively run within the trusted transactions.
-- In contrast, using the `RandomBeaconHistory` means to use a deterministically-seeded PRNG. 
-  The `RandomBeaconHistory` is key for effectively implementing a commit-and-reveal scheme.
-  During the commit phase, the user commits to proceed with a future source of randomness,
-  which is revealed after the commit transaction concluded.
-  For each block, the `RandomBeaconHistory` automatically stores the subsequently generated source of randomness.
-
-> 🚨 A transaction can atomically revert all its action during its runtime and abort.
+- Under the hood, the FVM instantiates a PRNG for each transaction from which `revertibleRandom` generates random numbers. 
+  However, when using `revertibleRandom` a developer is relying on the PRNG that is controlled by the transaction,
```
[the remark this answers is not in the document]
Comment (sisyphusSmiling) [reply README.md]: Updated in [5861871](https://github.com/onflow/random-coin-toss/pull/20/commits/58618718823a0ee680050799e70ead633250c538)

File: solidity/src/CadenceRandomConsumer.sol (hunk 1 above)
Comment (sisyphusSmiling) [reply solidity/src/CadenceRandomConsumer.sol:88]: Updated in [5861871](https://github.com/onflow/random-coin-toss/pull/20/commits/58618718823a0ee680050799e70ead633250c538)

File: solidity/src/CoinToss.sol [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        require(hasOpenRequest(msg.sender), "Caller has not flipped a coin - nothing to reveal");
+
+        // reveal random result and calculate winnings
+        uint256 requestId = coinTosses[msg.sender];
+        // delete the open request from the coinTosses mapping
+        delete coinTosses[msg.sender];
+
+        // fulfill the random request
+        uint64 randomResult = _fulfillRandomness(uint32(requestId));
+        uint8 coinFace = uint8(randomResult % 2);
```
Comment (sisyphusSmiling) [inline solidity/src/CoinToss.sol:59]: @tarakby is it safe to use modulo here to represent a resulting heads or tails?

**Comment by sisyphusSmiling** [comment]
I was running into issues with `getRandomSource` calls to Cadence Arch, but have since fixed them and validated that the contract is operational on Testnet.

File: solidity/src/CoinToss.sol (hunk 3 above)
Comment (tarakby) [reply solidity/src/CoinToss.sol:59]: Yes the modulo use is safe when the modulo divides the initial random range. For instance here `2` divides `2^64`. In this case there won't be any bias

**Pull request merged**